### ISSUE TITLE
Handle systemd cgroups driver in docker runtime

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#28192]
Kubelet has a feature that lets users specify the parent cgroups hierarchy to use for docker containers - `--cgroup-root`. This flag is currently defaulted to "", which means that the runtime can choose to parent the containers under any cgroup of their choice.
#27204 is adding support for pod level cgroups and improving Quality Of Service. Kubelet will be managing cgroups as part of this feature. Kubelet requires runtimes to instruct pods to be nested under a specified cgroups hierarchy. Since cgroups naming format is not compatible between cgroups and systemd, this makes implementing #27204 harder.

Between kubelet and docker, there are four possible operating modes wrt. cgroups:
- Kubelet can use cgroupfs or systemd for managing cgroups
- Docker can use cgroupfs or systemd for managing cgroups

For a simple UX, we want kubelet to support all its features irrespective of how docker is configured. The default kubelet configuration should _just_ work with all supported configurations.
- To achieve that simple UX, we need kubelet to understand the cgroups driver that it has to use and that docker is using. 
- There is an existing proposal to add `--cgroups-driver` flag to kubelet to instruct kubelet to use cgroupfs or systemd. It is not clear what a sane default would be for this flag since recent versions of systemd prevent other processes from writing to cgroups.
- Kubelet and docker will have to use the same cgroups driver.
- As of docker versions >= v1.11, it is possible to introspect the cgroups driver that docker is using. On versions prior to that, how can kubelet detect the cgroups driver that docker is using?
  - One option is to use `--cgroup-driver` flag if it is not possible to infer the driver from docker. Given that the default docker configuration is to use `cgroupfs`, kubelet should also default to `cgroupfs`. If kubelet is asked to use systemd, then docker should also be configured to use systemd.
  - Another option is to declare that systemd cgroups driver is not supported for docker versions <= v1.10.x

The main reason for filing this issue is that #27204 requires kubelet to choose a default value for `--cgroup-root`. What default value should kubelet choose? 

Since #27204 is an opt-in feature, it is possible to require the user to specify a valid option for `--cgroup-root`, whenever they enable QOS Cgroups. 

To make this feature opt-out in the future, kubelet will have to choose a sane default though. To keep the UX simple, kubelet will have to ensure that its configuration matches that of docker. 
We can either have the user specify the cgroups driver to use or let Kubelet auto-detect it. I prefer the latter because it leads to one-less knob that the user has to understand.

The first milestone for #27204 does not involve supporting systemd. Our node e2e test framework does not support a heterogeneous testing environment. Since #27204 is the first feature that manages cgroups for docker containers, it is necessary to keep the node e2e testing environment homogeneous. 

As of now, excepting CoreOS, all other distros are using the cgroups driver with docker. To make progress on #27204, it is necessary to make the CoreOS image not use systemd. Once ConfigMaps support lands in kubelet #27980, it will be possible to support systemd cgroups driver for specific test images. cc @euank 

cc @derekwaynecarr @dchen1107 @dubstack 

Action items:
- [x] Convert CoreOS node e2e image to not use systemd
- [ ] Identify how to support docker versions <= v1.10.x (One option is to require the users to setup kubelet appropriately)
- [ ] Identify defaults for `--cgroup-root` flag for cgroupfs and systemd

**Comment by ncdc** [comment]
@vishh Fedora, Centos, and RHEL all use the systemd cgroups driver with docker OOTB. We really really need this to work with systemd from the start because of this.

**Comment by dubstack** [comment]
@ncdc We intend to make this work on systemd systems before it becomes an opt-out in v1.5. Ideally we would like to achieve this in the v1.4 itself. So its really important for us to achieve the first milestone as early as possible. Then we will work on adding support for it with systemd cgroup drivers for the second milestone. But in any case this will be an opt-in in v1.4. So we don't effect existing users in v1.4. Does that sound reasonable?

**Comment by ncdc** [comment]
As long as it doesn't break on systemd systems, which it sounds like you're saying, then no objections here. However, we will try to do everything we can to help make this work with systemd in v1.4. Please let us know how we can help. I really want to get this feature in with systemd support if at all possible.

**Comment by yifan-gu** [comment]
> As long as it doesn't break on systemd systems, which it sounds like you're saying, then no objections here. However, we will try to do everything we can to help make this work with systemd in v1.4. Please let us know how we can help. I really want to get this feature in with systemd support if at all possible.

+1

Which distro in node e2e doesn't support systemd? Shouldn't node e2e tests the latest stable distros ? @vishh

**Comment by euank** [comment]
Another option here would be to skip the test for now on CoreOS until the code can support a docker daemon running with the systemd cgroup driver.

If the goal of the node_e2e images is to ensure that these node features work on various distros, then that seems like the most honest way to represent that no, this feature is known not to work on that distro by default.

Is there a reason this isn't an option?

**Comment by dubstack** [comment]
@ncdc Yes we would not break existing users who do not opt-in for this feature in v1.4 . We add a flag to the kubelet for specifying if the user wants to enable the cgroup hierarchy. If the flag is not specified the system will behave as expected using both systemd and cgroupfs cgroup drivers.

Thanks for offering your help. We would like to have some systemd expertise when we start working on  adding support for systemd drivers. I have been coordinating with @derekwaynecarr. Will keep you in the loop when we start that work.

**Comment by vishh** [comment]
@euank 
> Another option here would be to skip the test for now on CoreOS until the code can support a docker daemon running with the systemd cgroup driver.

This is not possible with the node e2e infra today. Like I mentioned, @mtaufen is working on adding support for customized kubelet flags in testing. So its just a matter of time before we let distros have customized configuration.

> If the goal of the node_e2e images is to ensure that these node features work on various distros, then that seems like the most honest way to represent that this feature is known not to work on that distro by default.

It is the goal and we have not yet reached that goal yet. Feel free to help @mtaufen or improve the infrastructure in other means to reach this goal.

> Is there a reason this isn't an option?

We have at-least two goals here - Maintain feature velocity & Test distros as-is. I'm suggesting compromising on testing distros as-is temporarily to maintain feature velocity. This is only a temporary compromise though.

**Comment by vishh** [comment]
@yifan-gu 

> Which distro in node e2e doesn't support systemd? 

All other images excepting CoreOS don't use systemd cgroups driver. 

> Shouldn't node e2e tests the latest stable distros ? 

This is the goal. But, we do not have the necessary infrastructure changes in place to achieve this goal. Will CoreOS be able to help bridge the gap? As I mentioned another option is to run a separate test suite with customized configuration until upstream k8s can support configurable node e2e test infra.

**Comment by euank** [comment]
Issue ref'd above has a slight bit more discussion.

If we can, I'd rather be able to skip the test for a specific distro until we have a better solution (feature supports coreos, etc) so we don't lose coverage of a more-or-less-default CoreOS image.

It seems to me like the relative difference in work between skipping a distro and updating the coreos image with the overhead the current process has isn't actually that huge.

If you feel strongly that we should be switching CoreOS to cgroupfs temporarily, I won't block on this and we can just run with it.

**Comment by vishh** [comment]
@euank 

Skipping tests based on distro is not an option for this specific case. Since kubelet configuration is unified across all test nodes, once this feature is enabled via flags on a node using systemd cgroups driver, no pods will start. 
IMHO, the two options are:
1. Setup a per-node kubelet configuration (I think this makes the node e2e framework complex. #27980 is the best way to achieve this IMHO!)
2. Update CoreOS node to not use systemd driver temporarily.

**Comment by euank** [comment]
I think the right way to switch it would be to drop a file at `/etc/systemd/system/docker.service.d/override.conf` with

```
[Service]
Environment='DOCKER_CGROUPS=--exec-opt native.cgroupdriver=cgroupfs'
```

**Comment by vishh** [comment]
@euank We can use the same issue I think. When can we get a node with the overridden config? This change is blocking PRs.

**Comment by euank** [comment]
@yifan-gu was kind enough to offer to help here (thanks!), so quite soon.

**Comment by dubstack** [comment]
Thanks @yifan-gu :smile:

**Comment by yifan-gu** [comment]
@vishh New image created, you should be able to access it: https://console.cloud.google.com/compute/imagesDetail/global/images/test-e2e-node-coreos-alpha-1068-20160701-cgroupfs?project=coreos-gce-testing

Tests passed, and the cgroup driver is changed to cgroupfs

**Comment by derekwaynecarr** [comment]
I need to catch up on this issue more next week, but the Red Hat e2e setup will support testing with the systemd driver and would provide a heterogeneous environment.  I am able to assist in rounding out the systemd support per our previous chats during 1.4 release cycle. I would prefer we demonstrate the function works both with and without systemd before making it required in 1.5 to allow for soak time.  I will synch up with you guys early next week as usual.

**Comment by dims** [comment]
switching this to 1.6 as it's too late for 1.5. ok? (please switch it right back if you disagree)

**Comment by vishh** [comment]
@derekwaynecarr can we close this issue now? I suspect all the issues are addressed already.

**Comment by dchen1107** [comment]
I am closing this one. If you disagree, please feel free to reopen it or file a separate issue for follow up.
